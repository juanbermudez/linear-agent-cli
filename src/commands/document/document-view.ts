import { Command } from "@cliffy/command"
import { renderMarkdown } from "@littletof/charmd"
import { viewDocument } from "../../utils/linear.ts"
import { formatRelativeTime } from "../../utils/display.ts"
import { pipeToUserPager, shouldUsePager } from "../../utils/pager.ts"
import { bold, underline } from "@std/fmt/colors"
import { header, muted } from "../../utils/styling.ts"

interface ViewOptions {
  web?: boolean
  app?: boolean
  comments?: boolean
  pager?: boolean
  json?: boolean
  format?: string
}

export const viewCommand = new Command()
  .name("view")
  .description("View document details")
  .arguments("<docId:string>")
  .option("-w, --web", "Open in web browser")
  .option("-a, --app", "Open in Linear.app")
  .option("--no-comments", "Exclude comments from output")
  .option("--no-color", "Disable colored output")
  .option("--no-pager", "Disable automatic paging")
  .option("-j, --json", "Output result as JSON (for AI agents)")
  .option("--format <format:string>", "Output format: text|json")
  .action(async (options: ViewOptions, docId: string) => {
    const useJson = options.json || options.format === "json"
    const showComments = options.comments !== false
    const usePager = options.pager !== false

    // TODO: Add web/app open support once we have the URL pattern
    if (options.web || options.app) {
      console.log("Opening document in browser/app not yet implemented")
      return
    }

    // Fetch document
    const { Spinner } = await import("@std/cli/unstable-spinner")
    const showSpinner = !useJson && Deno.stdout.isTerminal()
    const spinner = showSpinner
      ? new Spinner({ message: `Fetching document ${docId}...` })
      : null

    spinner?.start()

    try {
      const document = await viewDocument(docId)

      spinner?.stop()

      // Handle JSON output
      if (useJson) {
        console.log(
          JSON.stringify(
            {
              success: true,
              operation: "view",
              document: {
                id: document.id,
                title: document.title,
                slugId: document.slugId,
                url: document.url,
                icon: document.icon,
                color: document.color,
                content: document.content,
                creator: document.creator
                  ? {
                    id: document.creator.id,
                    name: document.creator.displayName,
                  }
                  : null,
                updatedBy: document.updatedBy
                  ? {
                    id: document.updatedBy.id,
                    name: document.updatedBy.displayName,
                  }
                  : null,
                project: document.project
                  ? {
                    id: document.project.id,
                    name: document.project.name,
                    url: document.project.url,
                  }
                  : null,
                initiative: document.initiative
                  ? {
                    id: document.initiative.id,
                    name: document.initiative.name,
                    url: document.initiative.url,
                  }
                  : null,
                createdAt: document.createdAt,
                updatedAt: document.updatedAt,
                comments: showComments
                  ? document.comments.nodes.map((comment) => ({
                    id: comment.id,
                    body: comment.body,
                    createdAt: comment.createdAt,
                    user: comment.user
                      ? {
                        id: comment.user.id,
                        name: comment.user.displayName,
                      }
                      : null,
                  }))
                  : undefined,
              },
            },
            null,
            2,
          ),
        )
        return
      }

      // Text output - render markdown
      const outputLines: string[] = []

      // Title with icon
      const titleLine = document.icon
        ? `${document.icon} ${document.title}`
        : document.title
      outputLines.push(bold(underline(titleLine)))
      outputLines.push("")

      // Metadata
      outputLines.push(
        muted(
          `Created by ${document.creator?.displayName || "Unknown"} ${
            formatRelativeTime(document.createdAt)
          }`,
        ),
      )
      if (
        document.updatedBy && document.creator &&
        document.updatedBy.id !== document.creator.id
      ) {
        outputLines.push(
          muted(
            `Updated by ${document.updatedBy.displayName} ${
              formatRelativeTime(document.updatedAt)
            }`,
          ),
        )
      } else {
        outputLines.push(
          muted(`Updated ${formatRelativeTime(document.updatedAt)}`),
        )
      }

      if (document.project) {
        outputLines.push(muted(`Project: ${document.project.name}`))
      }
      if (document.initiative) {
        outputLines.push(muted(`Initiative: ${document.initiative.name}`))
      }

      outputLines.push("")
      outputLines.push(document.url)
      outputLines.push("")

      // Content
      if (document.content && document.content.trim()) {
        if (Deno.stdout.isTerminal()) {
          let terminalWidth = 120
          try {
            const size = Deno.consoleSize()
            terminalWidth = size.columns
          } catch {
            // Use default
          }

          const renderedMarkdown = renderMarkdown(document.content, {
            lineWidth: terminalWidth,
          })
          outputLines.push(...renderedMarkdown.split("\n"))
        } else {
          outputLines.push(document.content)
        }
      } else {
        outputLines.push(muted("(No content)"))
      }

      // Comments
      if (showComments && document.comments.nodes.length > 0) {
        outputLines.push("")
        outputLines.push("")
        outputLines.push(header(`Comments (${document.comments.nodes.length})`))
        outputLines.push("")

        for (const comment of document.comments.nodes) {
          outputLines.push(
            bold(
              `${comment.user?.displayName || "Unknown"} ${
                muted(formatRelativeTime(comment.createdAt))
              }`,
            ),
          )

          if (Deno.stdout.isTerminal()) {
            let terminalWidth = 120
            try {
              const size = Deno.consoleSize()
              terminalWidth = size.columns
            } catch {
              // Use default
            }

            const renderedComment = renderMarkdown(comment.body, {
              lineWidth: terminalWidth,
            })
            outputLines.push(...renderedComment.split("\n"))
          } else {
            outputLines.push(comment.body)
          }

          outputLines.push("")
        }
      }

      // Output with paging if needed
      const output = outputLines.join("\n")
      if (usePager && shouldUsePager(outputLines, true)) {
        await pipeToUserPager(output)
      } else {
        console.log(output)
      }
    } catch (err) {
      spinner?.stop()

      const errorMsg = (err as Error).message.includes("not found")
        ? `Document '${docId}' not found`
        : `Failed to fetch document: ${(err as Error).message}`

      if (useJson) {
        console.error(
          JSON.stringify(
            {
              success: false,
              error: {
                code: (err as Error).message.includes("not found")
                  ? "NOT_FOUND"
                  : "API_ERROR",
                message: errorMsg,
                ...((err as Error).message.includes("not found") && {
                  resource: "document",
                  id: docId,
                }),
              },
            },
            null,
            2,
          ),
        )
      } else {
        console.error(`Error: ${errorMsg}`)
      }
      Deno.exit(1)
    }
  })
