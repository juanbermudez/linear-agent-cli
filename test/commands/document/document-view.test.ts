import { snapshotTest } from "@cliffy/testing"
import { viewCommand } from "../../../src/commands/document/document-view.ts"
import {
  commonDenoArgs,
  setupMockLinearServer,
} from "../../utils/test-helpers.ts"

// Test help output
await snapshotTest({
  name: "Document View Command - Help Text",
  meta: import.meta,
  colors: false,
  args: ["--help"],
  denoArgs: commonDenoArgs,
  async fn() {
    await viewCommand.parse()
  },
})

// Test viewing a document with markdown rendering
await snapshotTest({
  name: "Document View Command - Markdown Rendering",
  meta: import.meta,
  colors: false,
  args: ["doc-123", "--plain"],
  denoArgs: commonDenoArgs,
  async fn() {
    const { cleanup } = await setupMockLinearServer([
      {
        queryName: "GetDocument",
        variables: { id: "doc-123" },
        response: {
          data: {
            document: {
              id: "doc-123",
              title: "API Design Document",
              slugId: "api-design-document",
              url: "https://linear.app/test/document/api-design-document",
              content:
                "# API Design\n\nThis document describes the API design for the new backend service.\n\n## Overview\n\nThe API will use REST principles with JSON payloads.",
              project: {
                id: "project-789",
                name: "Backend Platform",
                slugId: "backend-platform",
              },
              creator: {
                name: "alice",
                displayName: "Alice Smith",
              },
              createdAt: "2024-01-10T10:00:00Z",
              updatedAt: "2024-01-15T14:30:00Z",
            },
          },
        },
      },
    ])

    try {
      await viewCommand.parse()
    } finally {
      await cleanup()
    }
  },
})

// Test viewing a document with JSON output
await snapshotTest({
  name: "Document View Command - JSON Output",
  meta: import.meta,
  colors: false,
  args: ["doc-456", "--json", "--plain"],
  denoArgs: commonDenoArgs,
  async fn() {
    const { cleanup } = await setupMockLinearServer([
      {
        queryName: "GetDocument",
        variables: { id: "doc-456" },
        response: {
          data: {
            document: {
              id: "doc-456",
              title: "Meeting Notes",
              slugId: "meeting-notes",
              url: "https://linear.app/test/document/meeting-notes",
              content:
                "# Meeting Notes\n\n- Discussed API design\n- Next steps: prototype",
              project: null,
              creator: {
                name: "bob",
                displayName: "Bob Jones",
              },
              createdAt: "2024-01-12T09:00:00Z",
              updatedAt: "2024-01-12T11:30:00Z",
            },
          },
        },
      },
    ])

    try {
      await viewCommand.parse()
    } finally {
      await cleanup()
    }
  },
})

// Test viewing a document without project linkage
await snapshotTest({
  name: "Document View Command - No Project Linked",
  meta: import.meta,
  colors: false,
  args: ["doc-solo", "--plain"],
  denoArgs: commonDenoArgs,
  async fn() {
    const { cleanup } = await setupMockLinearServer([
      {
        queryName: "GetDocument",
        variables: { id: "doc-solo" },
        response: {
          data: {
            document: {
              id: "doc-solo",
              title: "Standalone Document",
              slugId: "standalone-document",
              url: "https://linear.app/test/document/standalone-document",
              content:
                "# Standalone\n\nThis document is not linked to any project.",
              project: null,
              creator: {
                name: "charlie",
                displayName: "Charlie Brown",
              },
              createdAt: "2024-01-14T15:00:00Z",
              updatedAt: "2024-01-14T15:00:00Z",
            },
          },
        },
      },
    ])

    try {
      await viewCommand.parse()
    } finally {
      await cleanup()
    }
  },
})

// Test viewing a document with empty content
await snapshotTest({
  name: "Document View Command - Empty Content",
  meta: import.meta,
  colors: false,
  args: ["doc-empty", "--plain"],
  denoArgs: commonDenoArgs,
  async fn() {
    const { cleanup } = await setupMockLinearServer([
      {
        queryName: "GetDocument",
        variables: { id: "doc-empty" },
        response: {
          data: {
            document: {
              id: "doc-empty",
              title: "New Document",
              slugId: "new-document",
              url: "https://linear.app/test/document/new-document",
              content: null,
              project: {
                id: "project-123",
                name: "Mobile App",
                slugId: "mobile-app",
              },
              creator: {
                name: "alice",
                displayName: "Alice Smith",
              },
              createdAt: "2024-01-16T08:00:00Z",
              updatedAt: "2024-01-16T08:00:00Z",
            },
          },
        },
      },
    ])

    try {
      await viewCommand.parse()
    } finally {
      await cleanup()
    }
  },
})

// Test error when document not found
await snapshotTest({
  name: "Document View Command - Not Found",
  meta: import.meta,
  colors: false,
  args: ["doc-nonexistent", "--json", "--plain"],
  denoArgs: commonDenoArgs,
  async fn() {
    const { cleanup } = await setupMockLinearServer([
      {
        queryName: "GetDocument",
        variables: { id: "doc-nonexistent" },
        response: {
          data: {
            document: null,
          },
        },
      },
    ])

    try {
      await viewCommand.parse()
    } catch (_error) {
      // Expected to fail - document not found
    } finally {
      await cleanup()
    }
  },
})

// Test viewing document by slug ID
await snapshotTest({
  name: "Document View Command - By Slug ID",
  meta: import.meta,
  colors: false,
  args: ["api-design-document", "--plain"],
  denoArgs: commonDenoArgs,
  async fn() {
    const { cleanup } = await setupMockLinearServer([
      {
        queryName: "GetDocument",
        variables: { id: "api-design-document" },
        response: {
          data: {
            document: {
              id: "doc-789",
              title: "API Design Document",
              slugId: "api-design-document",
              url: "https://linear.app/test/document/api-design-document",
              content:
                "# API Design\n\nDetailed specifications for the new API.",
              project: {
                id: "project-456",
                name: "Backend Services",
                slugId: "backend-services",
              },
              creator: {
                name: "alice",
                displayName: "Alice Smith",
              },
              createdAt: "2024-01-10T10:00:00Z",
              updatedAt: "2024-01-15T14:30:00Z",
            },
          },
        },
      },
    ])

    try {
      await viewCommand.parse()
    } finally {
      await cleanup()
    }
  },
})
