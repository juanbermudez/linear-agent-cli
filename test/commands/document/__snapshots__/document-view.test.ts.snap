export const snapshot = {};

snapshot[`Document View Command - Help Text 1`] = `
stdout:
"
Usage: view <docId>

Description:

  View document details

Options:

  -h, --help               - Show this help.                        
  -w, --web                - Open in web browser                    
  -a, --app                - Open in Linear.app                     
  --no-comments            - Exclude comments from output           
  --plain                  - Disable colored output                 
  --no-pager               - Disable automatic paging               
  -j, --json               - Output result as JSON (for AI agents)  
  --format       <format>  - Output format: text|json               

"
stderr:
""
`;
